train.py

```diff
@@ -0,0 +1,274 @@
+import uuid
+import os.path
+import itertools
+from typing import Callable
+
+import hydra
+import numpy as np
+import torch.optim
+import pandas as pd
+import torch.nn as nn
+import streamlit as st
+from tqdm import tqdm
+from fortepyan import MidiPiece
+from sklearn.metrics import f1_score
+from torch.utils.data import DataLoader
+from omegaconf import OmegaConf, DictConfig
+
+import wandb
+from model import PitchSeqNN
+from data.dataset import BagOfPitches
+from utils import test_model, piece_av_files
+
+
+@hydra.main(version_base=None, config_path="conf", config_name="config")
+def main(cfg: DictConfig):
+    classnames = ["Ludwig van Beethoven", "Franz Liszt"]
+    path = "models/ludwig-van-beethoven-franz-liszt.pt"
+    if os.path.isfile(path):
+        checkpoint = torch.load(path)
+        model = PitchSeqNN(cfg.model.hidden_layers, 128, 2)
+        model.load_state_dict(checkpoint["model_state_dict"])
+    else:
+        model = run_experiment(cfg, classnames=classnames)
```

Comment:
I think it's better to avoid mixes of model training and model loading like that. You can think about model "evaluation" as a completely separate process, that you only run after training is finished.

When using hydra, we can have separate configs for training and evaluation. Separate entry points can also be convenient:

```sh
# This will create and print a run id ...
python train.py composers=["bach", "chopin"]

# ... which we can use to run evaluations
python eval.py run_id="run id"
```

Regarding streamlit: I still don't have any best practices or usage patterns figured out for it, but my intuition is to also treat it as a separate procedure, that can use the `eval` module.

```py
# dashboard.py
import streamlit as st

from eval import some_eval

# Maybe st.combobox wiht available run ids?
checkpoint = load_checkpoint()
...
```
And then:
```sh
streamlit dashboard.py
```

With this approach you have 3 separate sub-applications to manage, and it should be easier to keep the logic clean and separated :)

Reply:
I was just thinking of doing that today! thanks!